RecoPixelVertexing/PixelTriplets/interface/HitTripletEDProducerT.h

```diff
@@ -0,0 +1,198 @@
+#ifndef RecoPixelVertexing_PixelTriplets_HitTripletEDProducerT_H
+#define RecoPixelVertexing_PixelTriplets_HitTripletEDProducerT_H
+
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "FWCore/Utilities/interface/RunningAverage.h"
+
+#include "RecoTracker/TkHitPairs/interface/IntermediateHitDoublets.h"
+#include "RecoTracker/TkHitPairs/interface/RegionsSeedingHitSets.h"
+#include "RecoPixelVertexing/PixelTriplets/interface/OrderedHitTriplets.h"
+#include "RecoPixelVertexing/PixelTriplets/interface/IntermediateHitTriplets.h"
+#include "RecoPixelVertexing/PixelTriplets/interface/LayerTriplets.h"
+
+template <typename T_HitTripletGenerator>
+class HitTripletEDProducerT: public edm::stream::EDProducer<> {
+public:
+  HitTripletEDProducerT(const edm::ParameterSet& iConfig);
+  ~HitTripletEDProducerT() = default;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  virtual void produce(edm::Event& iEvent, const edm::EventSetup& iSetup) override;
+
+private:
+  edm::EDGetTokenT<IntermediateHitDoublets> doubletToken_;
+
+  edm::RunningAverage localRA_;
+
+  T_HitTripletGenerator generator_;
+
+  const bool produceSeedingHitSets_;
+  const bool produceIntermediateHitTriplets_;
+};
+
+template <typename T_HitTripletGenerator>
+HitTripletEDProducerT<T_HitTripletGenerator>::HitTripletEDProducerT(const edm::ParameterSet& iConfig):
+  doubletToken_(consumes<IntermediateHitDoublets>(iConfig.getParameter<edm::InputTag>("doublets"))),
+  generator_(iConfig, consumesCollector()),
+  produceSeedingHitSets_(iConfig.getParameter<bool>("produceSeedingHitSets")),
+  produceIntermediateHitTriplets_(iConfig.getParameter<bool>("produceIntermediateHitTriplets"))
+{
+  if(!produceIntermediateHitTriplets_ && !produceSeedingHitSets_)
+    throw cms::Exception("Configuration") << "HitTripletEDProducerT requires either produceIntermediateHitTriplets or produceSeedingHitSets to be True. If neither are needed, just remove this module from your sequence/path as it doesn't do anything useful";
+  if(produceSeedingHitSets_)
+    produces<RegionsSeedingHitSets>();
+  if(produceIntermediateHitTriplets_)
+    produces<IntermediateHitTriplets>();
+}
+
+template <typename T_HitTripletGenerator>
+void HitTripletEDProducerT<T_HitTripletGenerator>::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("doublets", edm::InputTag("hitPairEDProducer"));
```

Comment:
Currently there is only one module that produces the input product (`HitPairEDProducer`). The `fillDescriptions()` of that generates a module with a label `hitPairEDProducer`. I think the consistent naming of that module label and the default value of `doublets` parameter here serves as a good hint of the relationships of the modules.